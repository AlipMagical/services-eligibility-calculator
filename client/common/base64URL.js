/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const base64URLEncode = plainText =>
  btoa(plainText)
    .replace(/=/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_');

export const base64URLDecode = encodedText =>
  atob(encodedText.replace(/-/g, '+').replace(/_/g, '/'));

export default { base64URLEncode, base64URLDecode };
