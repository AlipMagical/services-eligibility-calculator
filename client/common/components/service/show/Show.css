/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.accepting-applications,
.not-accepting-applications,
.accepting-applications-future {
  display: inline-block;
  width: 100%;
  text-align: center;
  margin-top: 24px;
  border-radius: 8px;
}
.accepting-applications {
  color: #1e8e3e;
  background: #e6f4ea;
  border: 1px solid #a8dab5;
}
.not-accepting-applications {
  color: #5f6368;
}
.accepting-applications-future {
  background: #f1f3f4;
  color: #5f6368;
}

.apply-icon {
  margin-right: 20px;
}

.service-title {
  font-size: 28px;
  line-height: 36px;
  margin-bottom: 24px;
}

.material-icons {
  font-size: 20px;
}

.eligibility {
  align-items: center;
}

.eligibility .material-icons {
  padding: 5px;
  border-radius: 20px;
  color: #80868b;
  background: #f1f3f4;
}

.eligibility .col {
  margin-left: 16px;
}

.eligibility-other {
  margin-top: 7px;
}

.eligibility-or {
  text-align: center;
  border-bottom: 1px solid rgba(0, 0, 0, 0.125);
  line-height: 0;
  margin-bottom: 32px;
}

.eligibility-or span {
  background: white;
  padding: 0 10px;
}

.max-income .title {
  margin-top: 0px;
  margin-bottom: 15px;
  font-size: 32px;
  line-height: 40px;
}

.max-income .row {
  margin: 0px;
}

.max-income .col {
  padding: 0px;
}
.max-income hr {
  margin: 8px;
}
