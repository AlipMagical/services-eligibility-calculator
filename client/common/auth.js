/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { fetch } from './dataAccess';
import jwt_decode from 'jwt-decode';

export function login(username, password) {
  return fetch('login', {
    method: 'POST',
    body: JSON.stringify({
      username: username,
      password: password
    })
  }).then(handleTokenResponse);
}

export function refreshToken() {
  const refresh_token = localStorage.getItem('refresh_token');
  if (
    refresh_token !== null &&
    jwtWillExpireWithinSeconds(300 /* five minutes */)
  ) {
    return fetch('token/refresh', {
      method: 'POST',
      body: JSON.stringify({
        refresh_token
      })
    }).then(handleTokenResponse);
  }
}

function handleTokenResponse(response) {
  return response.json().then(json => {
    if (!json.token) {
      return false;
    }
    localStorage.setItem('jwt_token', json.token);
    if (json.refresh_token) {
      localStorage.setItem('refresh_token', json.refresh_token);
    } else {
      localStorage.removeItem('refresh_token');
    }
    return true;
  });
}

function jwtWillExpireWithinSeconds(seconds) {
  let encodedToken = localStorage.getItem('jwt_token');
  if (encodedToken != null) {
    let token = jwt_decode(encodedToken);
    return token.exp <= Math.floor(Date.now() / 1000) + seconds;
  }
  return false;
}

export function logout() {
  localStorage.removeItem('jwt_token');
  localStorage.removeItem('refresh_token');
}

export function isLoggedIn() {
  return getEncodedJwtToken() != null;
}

export function getEncodedJwtToken() {
  let encodedToken = localStorage.getItem('jwt_token');
  if (encodedToken != null) {
    let token = jwt_decode(encodedToken);
    if (token.exp <= Math.floor(Date.now() / 1000)) {
      localStorage.removeItem('jwt_token');
      encodedToken = null;
    }
  }
  return encodedToken;
}

export function getMillisUntilExpiration() {
  let encodedToken = getEncodedJwtToken();
  if (encodedToken != null) {
    let expiration = jwt_decode(encodedToken).exp * 1000;
    return expiration - Date.now();
  } else {
    return null;
  }
}

export function getLoggedInUser() {
  let encodedToken = getEncodedJwtToken();
  if (encodedToken != null) {
    return jwt_decode(encodedToken);
  } else {
    return {
      roles: [],
      username: '',
      displayName: '',
      mail: '',
      thumbnailPhoto: ''
    };
  }
}
