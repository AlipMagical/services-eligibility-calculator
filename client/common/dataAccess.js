/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ENTRYPOINT } from './config/entrypoint';
import get from 'lodash/get';
import { getEncodedJwtToken } from './auth';
import has from 'lodash/has';
import mapValues from 'lodash/mapValues';

const MIME_TYPE = 'application/ld+json';

export function fetch(url, options = {}) {
  if ('undefined' === typeof options.headers) options.headers = new Headers();
  if (null === options.headers.get('Accept'))
    options.headers.set('Accept', MIME_TYPE);

  if (
    'undefined' !== options.body &&
    !('undefined' !== typeof FormData && options.body instanceof FormData) &&
    null === options.headers.get('Content-Type')
  ) {
    options.headers.set('Content-Type', MIME_TYPE);
  }

  var jwtToken = getEncodedJwtToken();
  if (jwtToken) {
    options.headers.set('Authorization', 'Bearer ' + jwtToken);
  }

  var fullUrl = new URL(url, ENTRYPOINT);
  return global.fetch(fullUrl.href, options).then(response => {
    if (response.ok) return response;

    if (response.status === 401) {
      throw new Error(response.statusText || 'Authorization Error.');
    }

    return response.json().then(
      json => {
        const error =
          json['hydra:description'] ||
          json['hydra:title'] ||
          'An error occurred.';
        if (!json.violations) throw Error(error);

        let errors = { _error: error };
        json.violations.forEach(violation =>
          errors[violation.propertyPath]
            ? (errors[violation.propertyPath] +=
                '\n' + errors[violation.propertyPath])
            : (errors[violation.propertyPath] = violation.message)
        );

        throw new Error(errors);
      },
      () => {
        throw new Error(response.statusText || 'An error occurred.');
      }
    );
  });
}

export function mercureSubscribe(url, topics) {
  topics.forEach(topic =>
    url.searchParams.append('topic', new URL(topic, ENTRYPOINT))
  );

  return new EventSource(url.toString());
}

export function normalize(data) {
  if (has(data, 'hydra:member')) {
    // Normalize items in collections
    data['hydra:member'] = data['hydra:member'].map(item => normalize(item));

    return data;
  }

  // Flatten nested documents
  return mapValues(data, value =>
    Array.isArray(value)
      ? value.map(v => get(v, '@id', v))
      : get(value, '@id', value)
  );
}

export function extractHubURL(response) {
  const linkHeader = response.headers.get('Link');
  if (!linkHeader) return null;

  const matches = linkHeader.match(
    /<([^>]+)>;\s+rel=(?:mercure|"[^"]*mercure[^"]*")/
  );

  return matches && matches[1] ? new URL(matches[1], ENTRYPOINT) : null;
}
