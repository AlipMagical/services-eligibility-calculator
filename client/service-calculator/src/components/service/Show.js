/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import '../../common/components/service/show/Show.css';

import { Col, Row } from 'react-bootstrap';
import { ExternalLink, sendEmail } from '../../common/utils';
import React, { useContext } from 'react';

import ApplicationCard from '../../common/components/service/show/ApplicationCard';
import { BASEPATH } from '../../common/config';
import { DataContext } from '../../DataContext';
import DetailsCard from '../../common/components/service/show/DetailsCard';
import EligibilityCard from '../../common/components/service/show/EligibilityCard';
import { FilterContext } from '../../App';
import Helmet from 'react-helmet';
import OverviewCard from '../../common/components/service/show/OverviewCard';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import { TransparentButton } from '../eligibility-calculator/TransparentButton';
import { base64URLDecode } from '../../common/base64URL';
import { getTotalPeople } from '../../filters/filterAge';
import { useTranslation } from 'react-i18next';

export default function Show(props) {
  const serviceId = base64URLDecode(props.match.params.id);
  const { settings, servicesById } = useContext(DataContext);
  const service = servicesById && servicesById[serviceId];
  const { t } = useTranslation();
  const { filterState, lastSearch } = useContext(FilterContext);

  const shareViaEmail = () => {
    var emailBody = t('serviceDetails.shareViaEmailMessage');
    emailBody += '%0D%0A%0D%0A';
    emailBody += ((service && service.name) || '') + '%0D%0A';
    emailBody += encodeURIComponent(window.location.href);

    var emailSubject = t('serviceDetails.shareViaEmailSubject');
    sendEmail('', emailSubject, emailBody);

    ReactGA.event({
      category: 'Click',
      action: 'Share',
      label: 'Shared by Email'
    });
  };

  let resultsPage = BASEPATH;
  if (lastSearch) {
    resultsPage += '?q=' + encodeURIComponent(lastSearch);
  }

  return (
    <>
      {service === undefined && (
        <div role="status">
          <h2>{t('loading')}</h2>
        </div>
      )}

      {service && (
        <div className="service-details mx-auto">
          <Helmet>
            <title>{service.name}</title>
          </Helmet>
          <Row>
            <Col xs={12} className="mt-3">
              <TransparentButton
                className="d-xl-none"
                onClick={() => props.history.push(resultsPage)}
              >
                <i className="material-icons align-top" aria-hidden="true">
                  arrow_back
                </i>{' '}
                {t('serviceDetails.backButton')}
              </TransparentButton>
              <h1
                id="service-details-name"
                className="mt-1 mb-3 service"
                tabIndex="-1"
              >
                {service.name}
              </h1>
            </Col>
          </Row>
          <Row className="gutter-24">
            <Col xs={12} className="px-0 px-md-4">
              <OverviewCard service={service} hideTitle>
                <Row className="mt-4">
                  <Col xs={6}>
                    <button
                      className="btn btn-share w-100"
                      onClick={() => {
                        if (navigator.share) {
                          navigator
                            .share({
                              title: service.name,
                              text: service.name,
                              url: window.location.href
                            })
                            .then(() => {
                              ReactGA.event({
                                category: 'Click',
                                action: 'Share',
                                label: 'Native Share'
                              });
                            })
                            .catch(e => {
                              if (e.name === 'AbortError') {
                                // User cancelled share.
                                // Don't display sharing modal.
                                ReactGA.event({
                                  category: 'Click',
                                  action: 'Share',
                                  label: 'Cancelled'
                                });
                                return;
                              }
                              shareViaEmail();
                            });
                        } else {
                          shareViaEmail();
                        }
                      }}
                    >
                      {t('serviceDetails.share')}
                    </button>
                  </Col>
                  <Col xs={6}>
                    <ExternalLink
                      action="LearnMore"
                      url={service.applicationOnline}
                      className="btn btn-primary w-100"
                    >
                      {t('serviceDetails.learnMore')}
                    </ExternalLink>
                  </Col>
                </Row>
              </OverviewCard>
              <DetailsCard service={service} />
              <EligibilityCard
                eligibilityProfiles={service.eligibilityProfiles}
                ami={settings && settings.ami}
                fpl={settings && settings.fpl}
                householdSize={getTotalPeople(filterState.people)}
              />
              <ApplicationCard service={service} />
            </Col>
          </Row>
        </div>
      )}
    </>
  );
}

Show.propTypes = {
  match: PropTypes.object,
  onEdit: PropTypes.func
};
