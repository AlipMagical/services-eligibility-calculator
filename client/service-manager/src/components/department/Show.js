/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';

import { Modal } from 'react-bootstrap';
import PropTypes from 'prop-types';
import Remove from './Remove';
import { base64URLDecode } from '../../common/base64URL';
import { defaultLocale } from '../../Locales';
import { renderNonDefaultLocales } from '../namedTranslatedEntity';
import { useDepartmentTranslations } from '../../hooks';

export default function Show(props) {
  const departmentTranslations = useDepartmentTranslations(
    base64URLDecode(props.departmentId)
  );
  const [showRemoveModal, setShowRemoveModal] = useState(false);
  const department = departmentTranslations[defaultLocale];

  return (
    <>
      {showRemoveModal && (
        <Remove
          departmentId={props.departmentId}
          onCancel={() => {
            setShowRemoveModal(false);
          }}
          onRemove={() => {
            // Unwind the stack of modals.
            setShowRemoveModal(false);
            props.onCancel();
          }}
        />
      )}
      <Modal size="xl" show={showRemoveModal === false} onHide={props.onCancel}>
        <Modal.Body>
          <h1>Department: {department && department['name']}</h1>
          <>
            {department === undefined && (
              <div className="alert alert-info" role="status">
                Loading...
              </div>
            )}

            {department && (
              <table className="table table-responsive">
                <thead>
                  <tr>
                    <th>Field</th>
                    <th>Value</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <th scope="row">name</th>
                    <td>
                      {department['name']}{' '}
                      {renderNonDefaultLocales('name', departmentTranslations)}
                    </td>
                  </tr>
                </tbody>
              </table>
            )}
            <div className="float-left">
              <button onClick={props.onCancel} className="btn btn-blue-text">
                Close
              </button>
            </div>
            <div className="float-right">
              <button
                onClick={() => setShowRemoveModal(true)}
                className="btn btn-blue-text"
              >
                Remove
              </button>
              {
                <button onClick={props.onEdit} className="btn btn-blue-text">
                  Edit
                </button>
              }
            </div>
          </>
        </Modal.Body>
      </Modal>
    </>
  );
}

Show.propTypes = {
  departmentId: PropTypes.string,
  onCancel: PropTypes.func,
  onEdit: PropTypes.func
};
