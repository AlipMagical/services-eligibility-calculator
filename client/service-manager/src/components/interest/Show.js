/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';

import { Modal } from 'react-bootstrap';
import PropTypes from 'prop-types';
import Remove from './Remove';
import { base64URLDecode } from '../../common/base64URL';
import { defaultLocale } from '../../Locales';
import { renderNonDefaultLocales } from '../namedTranslatedEntity';
import { useInterestTranslations } from '../../hooks';

export default function Show(props) {
  const interestTranslations = useInterestTranslations(
    base64URLDecode(props.interestId)
  );
  const [showRemoveModal, setShowRemoveModal] = useState(false);
  const interest = interestTranslations[defaultLocale];

  return (
    <>
      {showRemoveModal && (
        <Remove
          interestId={props.interestId}
          onCancel={() => {
            setShowRemoveModal(false);
          }}
          onRemove={() => {
            // Unwind the stack of modals.
            setShowRemoveModal(false);
            props.onCancel();
          }}
        />
      )}
      <Modal size="xl" show={showRemoveModal === false} onHide={props.onCancel}>
        <Modal.Body>
          <h1>Interest: {interest && interest['name']}</h1>
          <>
            {interest === undefined && (
              <div className="alert alert-info" role="status">
                Loading...
              </div>
            )}
            {interest && (
              <table className="table table-responsive">
                <thead>
                  <tr>
                    <th>Field</th>
                    <th>Value</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <th scope="row">name</th>
                    <td>
                      {interest['name']}{' '}
                      {renderNonDefaultLocales('name', interestTranslations)}
                    </td>
                  </tr>
                  <tr>
                    <th scope="row">materialIcon</th>
                    <td>
                      <div className="row">
                        <div className="col-1">
                          <i className="material-icons">
                            {interest['materialIcon']}
                          </i>
                        </div>
                        <div className="col">{interest['materialIcon']}</div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            )}
            <div className="float-left">
              <button onClick={props.onCancel} className="btn btn-blue-text">
                Close
              </button>
            </div>
            <div className="float-right">
              <button
                onClick={() => setShowRemoveModal(true)}
                className="btn btn-blue-text"
              >
                Remove
              </button>
              {
                <button onClick={props.onEdit} className="btn btn-blue-text">
                  Edit
                </button>
              }
            </div>
          </>
        </Modal.Body>
      </Modal>
    </>
  );
}

Show.propTypes = {
  interestId: PropTypes.string,
  onCancel: PropTypes.func,
  onEdit: PropTypes.func
};
