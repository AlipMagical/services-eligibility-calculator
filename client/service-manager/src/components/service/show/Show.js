/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import '../../../common/components/service/show/Show.css';

import { AuthContext, SettingsContext } from '../../../App';
import { Col, Row } from 'react-bootstrap';
import React, { useContext, useEffect, useState } from 'react';
import {
  SERVICE_TYPE_ARCHIVED,
  SERVICE_TYPE_DRAFT,
  SERVICE_TYPE_HISTORY_VERSION,
  SERVICE_TYPE_PUBLISHED
} from '../../../common/utils/serviceTypes';
import { defaultLocale, locales } from '../../../Locales';

import AnalyticsCard from '../AnalyticsCard';
import ApplicationCard from '../../../common/components/service/show/ApplicationCard';
import { BASEPATH } from '../../../common/config';
import DetailsCard from '../../../common/components/service/show/DetailsCard';
import EligibilityCard from '../../../common/components/service/show/EligibilityCard';
import HistoryCard from '../HistoryCard';
import LanguageButtons from '../LanguageButtons';
import { Link } from 'react-router-dom';
import OverviewCard from '../../../common/components/service/show/OverviewCard';
import PropTypes from 'prop-types';
import Remove from '../Remove';
import ServiceOwnerCard from '../ServiceOwnerCard';
import Share from '../Share';
import StickyHeader from '../StickyHeader';
import { ToastContext } from '../../../App';
import { base64URLEncode } from '../../../common/base64URL';
import { getAllOwners } from '../../../common/utils';
import { useTranslation } from 'react-i18next';

/**
 * Displays information for a service.
 */
export default function Show({
  history,
  fromSettings,
  entityType,
  entityUrl,
  translations,
  serviceId,
  serviceHistory,
  editLink,
  locale,
  canManageService,
  onPublish
}) {
  const { settings } = useContext(SettingsContext);
  const [currentLocale, setCurrentLocale] = useState(locale || defaultLocale);
  const [entity, setEntity] = useState(translations[defaultLocale]);
  const { i18n } = useTranslation();
  useEffect(() => {
    i18n.changeLanguage(currentLocale);
    setEntity(translations[currentLocale]);
  }, [translations, currentLocale, i18n]);

  const [owners, setOwners] = useState(getAllOwners(entity));
  useEffect(() => {
    setOwners(getAllOwners(entity));
  }, [entity]);

  const [showRemoveModal, setShowRemoveModal] = useState(false);
  const [showShareModal, setShowShareModal] = useState(false);
  const { loggedInUser } = useContext(AuthContext);
  const { setToastMessage } = React.useContext(ToastContext);

  const editLinkInLocale = {
    ...editLink,
    state: { ...editLink.state, locale: currentLocale }
  };

  const renderActionButtons = () => {
    return (
      <>
        {entityType !== SERVICE_TYPE_HISTORY_VERSION &&
          canManageService(loggedInUser, entity) && (
            <div className="float-right">
              {entityType !== SERVICE_TYPE_ARCHIVED && (
                <>
                  <button
                    onClick={() => setShowRemoveModal(true)}
                    className="btn btn-blue-text"
                  >
                    {entityType === SERVICE_TYPE_DRAFT ? 'Delete' : 'Archive'}
                  </button>
                  <button
                    onClick={() => setShowShareModal(true)}
                    className="btn btn-blue-text"
                  >
                    Share
                  </button>
                </>
              )}
              {entityType === SERVICE_TYPE_DRAFT && (
                <>
                  <>
                    <Link
                      to={editLinkInLocale}
                      className="btn btn-blue-text"
                      role="button"
                    >
                      Edit
                    </Link>
                  </>
                  {onPublish && (
                    <button className="btn btn-primary" onClick={onPublish}>
                      Publish
                    </button>
                  )}
                </>
              )}
              {entityType === SERVICE_TYPE_PUBLISHED && (
                <Link
                  to={editLinkInLocale}
                  className="btn btn-primary"
                  role="button"
                >
                  Edit
                </Link>
              )}
              {entityType === SERVICE_TYPE_ARCHIVED && (
                <Link
                  to={editLinkInLocale}
                  className="btn btn-primary"
                  role="button"
                >
                  Restore
                </Link>
              )}
            </div>
          )}
        {entityType === SERVICE_TYPE_HISTORY_VERSION && (
          <div className="float-right">
            {canManageService(loggedInUser, entity) && (
              <Link to={editLinkInLocale}>
                <button className="btn btn-blue-text">Restore</button>
              </Link>
            )}
            <Link
              to={`${BASEPATH}services/show/${base64URLEncode(entity.service)}`}
            >
              <button className="btn btn-primary">Show Current Version</button>
            </Link>
          </div>
        )}
      </>
    );
  };

  return (
    <>
      {showRemoveModal && (
        <Remove
          serviceUrl={entityUrl}
          onCancel={() => setShowRemoveModal(false)}
          onRemove={() => {
            setShowRemoveModal(false);
            history.push(`${BASEPATH}services/`);
            setToastMessage(
              entity.name +
                (entityType === SERVICE_TYPE_DRAFT ? ' deleted' : ' archived')
            );
          }}
          draft={entityType === SERVICE_TYPE_DRAFT}
        />
      )}
      {showShareModal && (
        <Share
          serviceUrl={entityUrl}
          serviceName={entity.name}
          onCancel={() => setShowShareModal(false)}
          onShared={sharedWith => {
            setShowShareModal(false);
            setOwners([...owners, sharedWith]);
          }}
        />
      )}
      {entity === undefined && (
        <div className="alert alert-info" role="status">
          Loading...
        </div>
      )}

      {entity && (
        <>
          <StickyHeader
            title={entity.name}
            serviceType={entityType}
            backLink={
              fromSettings
                ? `${BASEPATH}settings/service-activity`
                : `${BASEPATH}services/`
            }
            actionButtons={renderActionButtons()}
          />
          <LanguageButtons
            locales={locales}
            locale={currentLocale}
            onLocaleChange={setCurrentLocale}
          />
          <Row>
            <Col xs={12} md={8} className="px-0 px-md-3">
              <OverviewCard service={entity} />
              <DetailsCard service={entity} />
              <EligibilityCard
                eligibilityProfiles={entity.eligibilityProfiles}
                ami={settings && settings.ami}
                fpl={settings && settings.fpl}
              />
              <ApplicationCard service={entity} />
            </Col>
            <Col xs={12} md={4} className="px-0 px-md-3">
              <ServiceOwnerCard owners={owners} />
              <AnalyticsCard serviceId={serviceId} />
              <HistoryCard
                history={serviceHistory}
                serviceId={serviceId}
                serviceType={entityType}
              />
            </Col>
          </Row>
        </>
      )}
    </>
  );
}

Show.propTypes = {
  history: PropTypes.object.isRequired,
  entityType: PropTypes.string.isRequired,
  entityUrl: PropTypes.string.isRequired,
  entity: PropTypes.object,
  serviceId: PropTypes.string,
  serviceHistory: PropTypes.array,
  editLink: PropTypes.object.isRequired,
  locale: PropTypes.string,
  canManageService: PropTypes.func.isRequired,
  onPublish: PropTypes.func
};
