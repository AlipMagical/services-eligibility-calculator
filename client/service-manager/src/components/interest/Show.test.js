/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { axe, toHaveNoViolations } from 'jest-axe';

import { MemoryRouter } from 'react-router';
import React from 'react';
import Show from './Show';
import { act } from 'react-dom/test-utils';
import { base64URLEncode } from '../../common/base64URL';
import { getModal } from '../../../../testing/testHelpers';
import { mockInterestJson } from '../../../../testing/mockData';
import { render } from 'react-dom';

expect.extend(toHaveNoViolations);

it('renders show an interest modal', async () => {
  const container = document.createElement('div');

  fetch.resetMocks();
  fetch.doMockOnceIf(
    url => url.includes('/api/interests/123'),
    mockInterestJson
  );
  fetch.doMockOnceIf(
    url => url.includes('/api/interests/123'),
    mockInterestJson
  );

  await act(async () => {
    return render(
      <MemoryRouter>
        <Show interestId={base64URLEncode('/api/interests/123')} />
      </MemoryRouter>,
      container
    );
  });

  expect(getModal()).toMatchSnapshot();
  expect(await axe(getModal())).toHaveNoViolations();
});
