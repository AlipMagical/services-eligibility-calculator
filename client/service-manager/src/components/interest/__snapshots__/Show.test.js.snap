// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders show an interest modal 1`] = `
<div
  aria-modal="true"
  class="fade modal show"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog modal-xl"
    role="document"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-body"
      >
        <h1>
          Interest: 
          Animals
        </h1>
        <table
          class="table table-responsive"
        >
          <thead>
            <tr>
              <th>
                Field
              </th>
              <th>
                Value
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                scope="row"
              >
                name
              </th>
              <td>
                Animals
                 
                (
                <i>
                  Spanish
                  : 
                  Animals
                </i>
                )
              </td>
            </tr>
            <tr>
              <th
                scope="row"
              >
                materialIcon
              </th>
              <td>
                <div
                  class="row"
                >
                  <div
                    class="col-1"
                  >
                    <i
                      class="material-icons"
                    >
                      pets
                    </i>
                  </div>
                  <div
                    class="col"
                  >
                    pets
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="float-left"
        >
          <button
            class="btn btn-blue-text"
          >
            Close
          </button>
        </div>
        <div
          class="float-right"
        >
          <button
            class="btn btn-blue-text"
          >
            Remove
          </button>
          <button
            class="btn btn-blue-text"
          >
            Edit
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
